File: string-paradigm-no-wrapper/src/main/java/com/steelzack/string/paradigm/nowrapper/performance/NoWrapperDependencyTreeModel.java
package com.steelzack.string.paradigm.nowrapper.performance;

import java.util.ArrayList;
import java.util.List;

import javax.swing.tree.DefaultMutableTreeNode;
import javax.swing.tree.DefaultTreeModel;
import javax.swing.tree.TreePath;

import com.steelzack.string.paradigm.api.Dependency;
import com.steelzack.string.paradigm.api.SegmentTreeModel;

public class NoWrapperDependencyTreeModel extends DefaultTreeModel implements SegmentTreeModel {

	/**
	 * 
	 */
	private static final long serialVersionUID = 7448300246415449363L;

	public NoWrapperDependencyTreeModel() {
		super(null);
		clear();
	}

	@Override
	public DefaultMutableTreeNode getRoot() {
		return (DefaultMutableTreeNode) super.getRoot();
	}

	@Override
	public DefaultMutableTreeNode getChild(Object parent, int index) {
		return (DefaultMutableTreeNode) super.getChild(parent, index);
	}

	@Override
	public void clear() {
		DefaultMutableTreeNode root = new DefaultMutableTreeNode("Combinations");
		root.setAllowsChildren(true);
		setRoot(root);
	}

	@Override
	public List<String> getDependencies(TreePath path) {
		DefaultMutableTreeNode last = (DefaultMutableTreeNode) path.getLastPathComponent();
		Object[] objects = last.getUserObjectPath();
		List<String> subDependencyElement = new ArrayList<String>();
		for (int i = 1; i < objects.length; i++) {
			subDependencyElement.add((String) objects[i]);
		}
		return subDependencyElement;
	}

	@Override
	public DefaultMutableTreeNode addCombination(Dependency dependency) {
		return addCombinationNode(getRoot(), dependency, 0);
	}

	@Override
	public DefaultMutableTreeNode addCombinationNode(DefaultMutableTreeNode parent, Dependency dependency,
			int codeIndex) {

		String subDependency = dependency.getSubDependencyElement(codeIndex);
		DefaultMutableTreeNode child = findChild(parent, subDependency);
		if (child == null) {
			child = new DefaultMutableTreeNode(subDependency);
			child.setAllowsChildren(true);
			insertNodeInto(child, parent, getChildCount(parent));
		}
		if (codeIndex < dependency.getDependencyCount() - 1) {
			addCombinationNode(child, dependency, codeIndex + 1);
		}
		return child;
	}

	@Override
	public DefaultMutableTreeNode findChild(DefaultMutableTreeNode parent, String subDependency) {
		for (int i = 0; i < getChildCount(parent); i++) {
			DefaultMutableTreeNode child = getChild(parent, i);
			if (child.getUserObject().equals(subDependency)) {
				return child;
			}
		}
		return null;
	}
}

File: string-paradigm-no-wrapper/src/main/kotlin/org/jesperancinha/string/paradigm/nowrapper/performance/NoWrapperDependencyTreeModel.kt
package org.jesperancinha.string.paradigm.nowrapper.performance

import org.jesperancinha.string.paradigm.api.Dependency
import org.jesperancinha.string.paradigm.api.SegmentTreeModel
import javax.swing.tree.DefaultMutableTreeNode
import javax.swing.tree.DefaultTreeModel
import javax.swing.tree.TreePath

class NoWrapperDependencyTreeModel : DefaultTreeModel(null), SegmentTreeModel {
    init {
        clear()
    }

    override fun getRoot(): DefaultMutableTreeNode {
        return super.getRoot() as DefaultMutableTreeNode
    }

    override fun getChild(parent: Any, index: Int): DefaultMutableTreeNode {
        return super.getChild(parent, index) as DefaultMutableTreeNode
    }

    override fun clear() {
        val root = DefaultMutableTreeNode("Combinations")
        root.allowsChildren = true
        setRoot(root)
    }

    override fun getDependencies(path: TreePath): List<String> {
        val last: DefaultMutableTreeNode = path.lastPathComponent as DefaultMutableTreeNode
        val objects: Array<Any> = last.userObjectPath
        val subDependencyElement: MutableList<String> = ArrayList()
        for (i in 1 until objects.size) {
            subDependencyElement.add(objects[i] as String)
        }
        return subDependencyElement
    }

    override fun addCombination(dependency: Dependency) = addCombinationNode(getRoot(), dependency, 0)

    override fun addCombinationNode(
        parent: DefaultMutableTreeNode,
        dependency: Dependency,
        codeIndex: Int
    ): DefaultMutableTreeNode {
        val subDependency = dependency.getSubDependencyElement(codeIndex)
        var child: DefaultMutableTreeNode? = parent.let {
            subDependency?.let { it1 ->
                try {
                    findChild(it, it1)
                } catch (_: Exception) {
                    null
                }
            }
        }
        if (child == null) {
            child = DefaultMutableTreeNode(subDependency)
            child.allowsChildren = true
            insertNodeInto(child, parent, getChildCount(parent))
        }
        if (codeIndex < dependency.dependencyCount() - 1) {
            addCombinationNode(child, dependency, codeIndex + 1)
        }
        return child
    }

    override fun findChild(parent: DefaultMutableTreeNode, subDependency: String): DefaultMutableTreeNode {
        for (i in 0 until getChildCount(parent)) {
            val child: DefaultMutableTreeNode = getChild(parent, i)
            if (child.userObject == subDependency) {
                return child
            }
        }
        throw NullPointerException()
    }

    companion object {
        /**
         *
         */
        private const val serialVersionUID = 7448300246415449363L
    }
}